File: celladapter/src/main/java/io/erva/celladapter/v7/Cell.java
package io.erva.celladapter.v7;

import android.support.v7.widget.RecyclerView;
import android.view.View;

public abstract class Cell<ITEM, LISTENER extends Cell.Listener<ITEM>> extends RecyclerView.ViewHolder {

    private ITEM item;
    private LISTENER listener;

    public Cell(View view) {
        super(view);
        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (listener != null) listener.onCellClicked(getItem());
            }
        });
        view.addOnAttachStateChangeListener(new View.OnAttachStateChangeListener() {
            @Override
            public void onViewAttachedToWindow(View v) {
            }

            @Override
            public void onViewDetachedFromWindow(View v) {
                clearResources();
                v.removeOnAttachStateChangeListener(this);
            }
        });
    }

    protected abstract void bindView();

    protected final ITEM getItem() {
        return item;
    }

    protected LISTENER getListener() {
        return listener;
    }

    protected void clearResources() {
    }

    void setCellDelegate(LISTENER listener) {
        this.listener = listener;
    }

    void bindViewInternal(ITEM item) {
        this.item = item;
        bindView();
    }

    public interface Listener<ITEM> {

        void onCellClicked(ITEM item);
    }
}

File: celladapter-kotlin/src/main/kotlin/io/erva/celladapter/v7/Cell.kt
package io.erva.celladapter.v7

import android.support.v7.widget.RecyclerView
import android.view.View

abstract class Cell<ITEM : Any, out LISTENER : Cell.Listener<ITEM>>(val view: View) : RecyclerView.ViewHolder(view) {

    var item: Any? = null
        set(value) {
            field = value
            bindView()
        }
    private var listener: Any? = null

    init {
        view.setOnClickListener { listener()?.onCellClicked(item()) }
    }

    abstract fun bindView()

    fun item(): ITEM {
        return item as ITEM
    }

    fun item(item: Any) {
        this.item = item
    }

    fun listener(): LISTENER? {
        return listener as LISTENER?
    }

    fun listener(listener: Any?) {
        this.listener = listener
    }

    interface Listener<ITEM : Any> {

        fun onCellClicked(item: ITEM) {
        }
    }
}